src/io_scheduler.cpp

```diff
@@ -65,21 +58,29 @@ io_scheduler::~io_scheduler()
         m_io_thread.join();
     }
 
-    if (m_epoll_fd != -1)
+    if (m_shutdown_fd[0] != -1)
     {
-        close(m_epoll_fd);
-        m_epoll_fd = -1;
+        close(m_shutdown_fd[0]);
+        m_shutdown_fd[0] = -1;
     }
-    if (m_timer_fd != -1)
+    if (m_shutdown_fd[1] != -1)
+    {
+        close(m_shutdown_fd[1]);
+        m_shutdown_fd[1] = -1;
+    }
+
+    if (m_schedule_fd[0] != -1)
     {
-        close(m_timer_fd);
-        m_timer_fd = -1;
+        close(m_schedule_fd[0]);
+        m_schedule_fd[0] = -1;
     }
-    if (m_schedule_fd != -1)
+    if (m_schedule_fd[1] != -1)
     {
-        close(m_schedule_fd);
-        m_schedule_fd = -1;
+        close(m_schedule_fd[1]);
+        m_schedule_fd[1] = -1;
     }
+
+    // TODO Close all pipe-fds
```

Comment:
are these all handled?

Reply:
Thanks.

I'm looking into some of the failures too, I _think_ they are unrelated, but I'm also having a hard time reproducing locally. I've been running the two fedora failures locally for about an hour and cannot get it to trigger :grimacing:

Reply:
> Great! I appreciate your help with that :)
> 
> I figured that I can only trigger the segfaults _without_ sanitizers enabled on my Mac but I will keep looking into that. I really want to get this merged after all the work at some point :D

Totally, its a very nice change. I'm probably going to open bugs for the fedora/conan issues and track those separately is my current though.

Reply:
It appears it might be indeed. I can merge that portion of the fix and we can see if this dissapears. There is another problem when a `when_any` test as well with a bad pointer free/segfault I'm still working through, but let me split that off and get the 270 fix re-merged.

Reply:
Oh perfect, it seems to fix the segfault and timeout problems for macOS completely! I will rebase the commits into one commit and check if I need to make changes for the static code analysis job to complete and then this should be ready. Thanks a lot!